Create 2022.01 release

## Summary

In order to improve the acceptance of Gitpod self-hosted by the community and enterprise users, a formalised and regular release cycle of the Gitpod Installer is critical. This will be the first release of the Installer binaries and will be the template for how the process will run in future.

## Context

Currently, the process is to instruct the community to find a specific (and arbitrary) build in Werft and use the build ID from there.

## Acceptance Criteria

A release of the Installer is made with binaries for all the supported architectures

## In Scope

 - Builds of the Go binary for the following os/architectures (see `go tool dist list`):
   - linux/amd64
   - linux/386
   - linux/arm
   - linux/arm64
   - ~darwin/amd64~
   - ~darwin/arm64~
   - ~windows/386~
   - ~windows/amd64~
 - Creation of a `2022.01` release in GitHub releases
 - Creation of a `2022.01` tag
 - Update the Gitpod guides to use the `2022.01` release
 - Tested and working with in-cluster and external dependencies in:
   - AWS
   - Azure
   - GCP

## Out of Scope

 - Automation of pipeline. This is about working out the desired process. Automation will come for February's `2022.02` release
 - Use of Replicated
 - Non-Linux machines:
   - darwin/amd64
   - darwin/arm64
   - windows/386
   - windows/amd64

## Tasks

- [x] #6766 
- [x] Create `releases/2022.01` branch
- [ ] Create a `2022.01` tag
- [x] #7789
- [x] Build binaries and add to GitHub releases
- [ ] #7880
- [ ] Test on and update AWS guide
- [ ] Test on and update Azure guide
- [ ] Test on and update GCP guide
- [x] #7497
- [x] Update the Installer readme to remove references to `./installer` etc and replace with downloaded binary
- [ ] #7820

So, one of the plans we had was to build the Installer binary for both Mac and Windows. Seemingly, because [runc](https://github.com/opencontainers/runc/issues/1733) is only supported on Linux, when I run `GOOS=darwin GOARCH=amd64 go build`, we get a lot of `runc` and `containerd` compilation errors.

An excerpt
```shell
# github.com/opencontainers/runc/libcontainer/cgroups
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:79:14: undefined: unix.Openat2
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:79:45: undefined: unix.OpenHow
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:100:18: undefined: unix.RESOLVE_BENEATH
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:100:41: undefined: unix.RESOLVE_NO_MAGICLINKS
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:101:17: undefined: unix.CGROUP2_SUPER_MAGIC
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:103:20: undefined: unix.RESOLVE_NO_XDEV
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:103:43: undefined: unix.RESOLVE_NO_SYMLINKS
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/v1_utils.go:155:9: undefined: Mount
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/v1_utils.go:163:87: undefined: Mount
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/v1_utils.go:191:37: undefined: Mount
../../go/pkg/mod/github.com/opencontainers/runc@v1.0.1/libcontainer/cgroups/file.go:103:43: too many errors
```

What should our resolution be for this `2022.01` release? As I see it, the most straightforward option is to only support Linux for this release and offer the Docker script for Mac/Windows users.

Any other suggestions? @corneliusludmann @csweichel

--- a/installer/.gitignore
+++ b/installer/.gitignore
@@ -1,1 +1,3 @@
 installer
+dist
+versions.yaml

--- a/installer/Makefile
+++ b/installer/Makefile
@@ -1,0 +1,36 @@
+OUTPUT ?= dist
+VERSION_MANIFEST = ./pkg/config/versions/versions.yaml
+
+all: clean deps versionManifest compile
+
+build:
+	@echo "Building for ${GOOS}-${GOARCH}"
+	CGO_ENABLED=0 go build -tags embedVersion -o ${OUTPUT}/gitpod-installer-${GOOS}-${GOARCH}${EXT}
+	(cd ${OUTPUT} && sha256sum gitpod-installer-${GOOS}-${GOARCH} > gitpod-installer-${GOOS}-${GOARCH}.sha256)
+.PHONY: build
+
+clean:
+	@echo "Cleaning ${OUTPUT}"
+	rm -Rf ${OUTPUT}
+	mkdir -p ${OUTPUT}
+	rm -f ${VERSION_MANIFEST}
+.PHONY: clean
+
+compile:
+	@echo "Compiling for every supported OS and platform"
+
+	GOOS=linux GOARCH=amd64 $(MAKE) build
+	GOOS=linux GOARCH=386 $(MAKE) build
+	GOOS=linux GOARCH=arm $(MAKE) build
+	GOOS=linux GOARCH=arm64 $(MAKE) build
+.PHONY: compile
+
+deps:
+	@echo "Installing Helm dependencies"
+	@for f in $(shell ls -d third_party/charts/*/); do cd $${f} && helm dep up && cd -; done
+.PHONY: deps
+
+versionManifest:
+	@echo "Downloading version manifest for ${VERSION}"
+	docker run -it --rm eu.gcr.io/gitpod-core-dev/build/versions:${VERSION} cat versions.yaml > ${VERSION_MANIFEST}
+.PHONY: versionManifest

--- a/installer/README.md
+++ b/installer/README.md
@@ -14,8 +14,8 @@ The best way to get started with Gitpod
 
 # Requirements
 
-- A machine running Linux/MacOS and Docker
-    - Windows is not currently supported, but will be in future
+- A machine running Linux
+    - MacOS and Windows are not currently supported, but may be in future
 - [Kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl) installed
 - A [Kubernetes cluster configured](https://www.gitpod.io/docs/self-hosted/latest/installation)
 - A [TLS certificate](#tls-certificates)
@@ -31,35 +31,53 @@ The process to install Gitpod is:
 
 # Quickstart
 
-## Get your binary
+## Download the Installer on Linux
 
-> There is work to be done on improving the distribution of the binary
-> [#6766](https://github.com/gitpod-io/gitpod/issues/6766).
+Releases can be downloaded from [GitHub Releases](https://github.com/gitpod-io/gitpod/releases/)
+Select your desired binary, download and install
 
-From [werft](https://werft.gitpod-dev.com), search for the tag you want, such as
-`main.1919` - set this value to `$TAG`
-
-### Option 1 - download the pre-build Installer
+1. Download the latest release with the command:
 
 ```shell
-docker create -ti --name installer eu.gcr.io/gitpod-core-dev/build/installer:$TAG
-docker cp installer:/app/installer ./installer
-docker rm -f installer
+curl -fsSLO https://github.com/gitpod-io/gitpod/releases/download/2022.01/gitpod-installer-linux-amd64
 ```
 
-### Option 2 - use `go run`
+2. Validate the binary (optional)
+
+Download the checksum file:
 
 ```shell
-# Get the versions
-docker run -it --rm eu.gcr.io/gitpod-core-dev/build/versions:$TAG cat /versions.yaml > versions.yaml
+curl -fsSLO https://github.com/gitpod-io/gitpod/releases/download/2022.01/gitpod-installer-linux-amd64.sha256
 ```
 
-For every command below, replace `installer` with `go run . --debug-version-file ./versions.yaml`.
+Validate the binary against the checksum file:
+
+```shell
+echo "$(<gitpod-installer-linux-amd64.sha256)" | sha256sum --check
+```
+
+If valid, the output is:
+
+```shell:
+gitpod-installer-linux-amd64: OK
+```
+
+3. Install the binary
+
+```shell
+sudo install -o root -g root gitpod-installer-linux-amd64 /usr/local/bin/gitpod-installer
+```
+
+4. Test to ensure the version you installed it up-to-date:
+
+```shell
+gitpod-installer version
+```
 
 ## Generate the base config
 
 ```shell
-./installer init > gitpod.config.yaml
+gitpod-installer init > gitpod.config.yaml
 ```
 
 ## Customise your config
@@ -74,7 +92,7 @@ own.
 
 ```shell
 # Checks the validity of the configuration YAML
-./installer validate config --config gitpod.config.yaml
+gitpod-installer validate config --config gitpod.config.yaml
 ```
 
 Any errors here must be fixed before deploying. See [Config](#config) for
@@ -82,7 +100,7 @@ more details.
 
 ```shell
 # Checks that your cluster is ready to install Gitpod
-./installer validate cluster --kubeconfig ~/.kube/config --config gitpod.config.yaml
+gitpod-installer validate cluster --kubeconfig ~/.kube/config --config gitpod.config.yaml
 ```
 
 Any errors here must be fixed before deploying. See [Cluster Dependencies](#cluster-dependencies)
@@ -91,7 +109,7 @@ for more details.
 ## Render the YAML
 
 ```shell
-./installer render --config gitpod.config.yaml > gitpod.yaml
+gitpod-installer render --config gitpod.config.yaml > gitpod.yaml
 ```
 
 ## Deploy
@@ -492,14 +510,14 @@ To install to a different namespace, pass a `namespace` flag to the `render`
 command.
 
 ```shell
-./installer render --config gitpod.config.yaml --namespace gitpod > gitpod.yaml
+gitpod-installer render --config gitpod.config.yaml --namespace gitpod > gitpod.yaml
 ```
 
 The `validate cluster` command also accepts a namespace, allowing you to run
 the checks on that namespace.
 
 ```shell
-./installer validate cluster --kubeconfig ~/.kube/config --config gitpod.config.yaml --namespace gitpod
+gitpod-installer validate cluster --kubeconfig ~/.kube/config --config gitpod.config.yaml --namespace gitpod
 ```
 
 **IMPORTANT**: this does not create the namespace, so you will need to create

--- a/installer/cmd/version.go
+++ b/installer/cmd/version.go
@@ -39,6 +39,14 @@ var versionCmd = &cobra.Command{
 }
 
 func getVersionManifest() (*versions.Manifest, error) {
+	embeddedManifest, err := versions.Embedded()
+	if err != nil {
+		return nil, err
+	}
+	if embeddedManifest != nil {
+		return embeddedManifest, nil
+	}
+
 	var data []byte
 	if rootOpts.VersionMF != "" {
 		var err error
@@ -68,7 +76,7 @@ func getVersionManifest() (*versions.Manifest, error) {
 	}
 
 	var res versions.Manifest
-	err := yaml.Unmarshal(data, &res)
+	err = yaml.Unmarshal(data, &res)
 	if err != nil {
 		return nil, err
 	}

--- a/installer/pkg/config/v1/config.go
+++ b/installer/pkg/config/v1/config.go
@@ -52,7 +52,7 @@ func (v version) Defaults(in interface{}) error {
 		corev1.ResourceCPU:    resource.MustParse("1000m"),
 		corev1.ResourceMemory: resource.MustParse("2Gi"),
 	}
-	cfg.Workspace.Runtime.FSShiftMethod = FSShiftFuseFS
+	cfg.Workspace.Runtime.FSShiftMethod = FSShiftShiftFS
 	cfg.Workspace.Runtime.ContainerDSocket = "/run/containerd/containerd.sock"
 	cfg.Workspace.Runtime.ContainerDRuntimeDir = "/var/lib/containerd/io.containerd.runtime.v2.task/k8s.io"
 

--- a/installer/pkg/config/versions/embed.go
+++ b/installer/pkg/config/versions/embed.go
@@ -1,0 +1,25 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+//go:build embedVersion
+
+package versions
+
+import (
+	_ "embed"
+
+	"sigs.k8s.io/yaml"
+)
+
+//go:embed versions.yaml
+var embeddedVersion []byte
+
+func loadEmbedded() (*Manifest, error) {
+	var res Manifest
+	err := yaml.Unmarshal(embeddedVersion, &res)
+	if err != nil {
+		return nil, err
+	}
+	return &res, nil
+}

--- a/installer/pkg/config/versions/noembed.go
+++ b/installer/pkg/config/versions/noembed.go
@@ -1,0 +1,11 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+//go:build !embedVersion
+
+package versions
+
+func loadEmbedded() (*Manifest, error) {
+	return nil, nil
+}

--- a/installer/pkg/config/versions/versions.go
+++ b/installer/pkg/config/versions/versions.go
@@ -62,3 +62,9 @@ type Components struct {
 	WSManagerBridge Versioned `json:"wsManagerBridge"`
 	WSProxy         Versioned `json:"wsProxy"`
 }
+
+// var embedded embed.FS
+
+func Embedded() (*Manifest, error) {
+	return loadEmbedded()
+}

